x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/AsyncRetryDuringSnapshotActionStep.java

```diff
@@ -0,0 +1,165 @@
+                              ClusterStateObserver observer, Listener listener) {
+        // Wrap the original listener to handle exceptions caused by ongoing snapshots
+        SnapshotExceptionListener snapshotExceptionListener = new SnapshotExceptionListener(indexMetaData.getIndex(), listener, observer);
+        performDuringNoSnapshot(indexMetaData, currentClusterState, snapshotExceptionListener);
+    }
+
+    /**
+     * Method to be performed during which no snapshots for the index are already underway.
+     */
+    abstract void performDuringNoSnapshot(IndexMetaData indexMetaData, ClusterState currentClusterState, Listener listener);
+
+    /**
+     * SnapshotExceptionListener is an injected listener wrapper that checks to see if a particular
+     * action failed due to a {@code SnapshotInProgressException}. If it did, then it registers a
+     * ClusterStateObserver listener waiting for the next time the snapshot is not running,
+     * re-running the step's {@link #performAction(IndexMetaData, ClusterState, ClusterStateObserver, Listener)}
+     * method when the snapshot is no longer running.
+     */
+    class SnapshotExceptionListener implements AsyncActionStep.Listener {
+        private final Index index;
+        private final Listener originalListener;
+        private final ClusterStateObserver observer;
+
+        SnapshotExceptionListener(Index index, Listener originalListener, ClusterStateObserver observer) {
+            this.index = index;
+            this.originalListener = originalListener;
+            this.observer = observer;
+        }
+
+        @Override
+        public void onResponse(boolean complete) {
+            originalListener.onResponse(complete);
+        }
+
+        @Override
+        public void onFailure(Exception e) {
+            if (e instanceof SnapshotInProgressException) {
+                try {
+                    logger.debug("[{}] attempted to run ILM step but a snapshot is in progress, step will retry at a later time",
+                        index.getName());
+                    observer.waitForNextChange(
+                        new NoSnapshotRunningListener(observer, index.getName(), state -> {
+                            IndexMetaData idxMeta = state.metaData().index(index);
+                            if (idxMeta == null) {
+                                // The index has since been deleted, mission accomplished!
+                                originalListener.onResponse(true);
+                            }
+                            // Re-invoke the performAction method with the new state
+                            performAction(idxMeta, state, observer, originalListener);
+                        }, originalListener::onFailure),
+                        // TODO: what is a good timeout value for no new state received during this time?
+                        TimeValue.timeValueHours(12));
```

Comment:
I think waiting 12 hours for a snapshot to finish is reasonable. If there is no progress on this action in that time interval, a user may want to know. so 👍